src/adapter/tracer/ptrace.rs

```diff
@@ -0,0 +1,453 @@
+    F: Fn(T, Duration) -> T,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let start = Instant::now();
+        let item = self.inner.next();
+        let time = start.elapsed();
+        match item {
+            Some(item) => Some((self.post_action)(item, time)),
+            None => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+fn make_iter_with_perf_span<I, T, F>(inner: I, post_action: F) -> PerfSpanIter<I, T, F>
+where
+    I: Iterator<Item = T>,
+    F: Fn(T, Duration) -> T,
+{
+    PerfSpanIter { inner, post_action }
+}
+
+/// Traces an inner adapter. Constructed with new(). Each tracer is only valid
+/// for one query. When switching between queries, ensure you first call finish(),
+/// otherwise the new operations will be traced with the old.
+#[derive(Debug, Clone)]
+pub struct TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex>,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    pub tracer: Rc<RefCell<Tracer>>,
+    inner: AdapterT,
+    _phantom: PhantomData<&'vertex ()>,
+}
+
+impl<'vertex, AdapterT> TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex>,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    pub fn new(adapter: AdapterT) -> Self {
+        Self {
+            tracer: Rc::new(RefCell::new(Tracer::new())),
+            inner: adapter,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Finalise the trace and return it.
+    pub fn finish(&self) -> Tracer {
+        // Ensure nothing is reading the trace i.e. we can safely stop interpreting.
+        let trace_ref = self.tracer.borrow_mut();
+        let new_trace = Tracer::new();
+        drop(trace_ref);
+        self.tracer.replace(new_trace)
+    }
+}
+
+impl<'vertex, AdapterT> Adapter<'vertex> for TracingAdapter<'vertex, AdapterT>
+where
+    AdapterT: Adapter<'vertex> + 'vertex,
+    AdapterT::Vertex: Debug + Clone + 'vertex,
+{
+    type Vertex = AdapterT::Vertex;
+
+    fn resolve_starting_vertices(
+        &self,
+        edge_name: &Arc<str>,
+        parameters: &EdgeParameters,
+        resolve_info: &ResolveInfo,
+    ) -> VertexIterator<'vertex, Self::Vertex> {
+        // We don't care about the first resolution since it's normally quick
+        // and it only occurs once.
+        self.inner
+            .resolve_starting_vertices(edge_name, parameters, resolve_info)
+    }
+
+    fn resolve_property<V: AsVertex<Self::Vertex> + 'vertex>(
+        &self,
+        contexts: ContextIterator<'vertex, V>,
+        type_name: &Arc<str>,
+        property_name: &Arc<str>,
+        resolve_info: &ResolveInfo,
+    ) -> ContextOutcomeIterator<'vertex, V, FieldValue> {
+        // For each resolution we want to know:
+        // 1. What are we resolving?
+        // 2. How long did the resolution take?
+        // 3. How many times did we resolve the same property?
```

Comment:
Clarification nit to avoid confusion, since in some cases the same property will be resolved more than once in the same query (e.g. `@filter @output` on the same property). That's not what this point implies we're tracking.
```suggestion
        // 3. For how many vertices did we resolve that property?
```

Reply:
To my mind, "how many vertices" reads as if we are counting the number of different vertices we resolved the property for, e.g. `Vid(1)` and `Vid(2)` and `Vid(3)`, not what we are counting which is number of calls to resolve a particular `(vid, type, property)` triple.

Reply:
I do wish we had better terminology. But inside the scope of this method, we only have the ability to observe a single `Vid(_)` and the statement is unambiguous. We also have to assume that the person reading this code has some familiarity with the Trustfall adapter API already, which would help disambiguate further.

> How many times did we resolve the same property?

This on the other hand is actively inaccurate. If a user's query includes `@filter @output` or `@output @output(name: "foo")`, that user would be justifiably displeased when they see this count is _not actually_ "how many times did we resolve the same property" in their query.

In the Trustfall API, the "resolve" action happens _once total per API call_ not once per data vertex whose property is being loaded. Think of "resolve" as the thing that happens when you call a function in an interpreted language like Python: the runtime has to _resolve_ the name of the function, and then call it. If that function has a loop inside, we don't say we _resolved_ it more than once. Trustfall similarly makes _one_ call into the adapter. That adapter is our function; if it internally chooses to make many function calls under the hood, that is an implementation detail.

This documentation has to work for any implementation of the Trustfall adapter API, so:
- we only resolved the property once
- we count for how many vertices that happened